lib/core/include/scipp/core/spatial_transforms.h

```diff
@@ -0,0 +1,139 @@
+    } else if (i == 3) {
+      return m_quat.w();
+    } else {
+      throw std::out_of_range("invalid index for Quaternion");
+    }
+  }
+};
+
+class Translation {
+private:
+  Eigen::Vector3d m_vec;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  Translation() : m_vec(Eigen::Vector3d(0, 0, 0)){};
+  explicit Translation(const Eigen::Vector3d &x) : m_vec(x){};
+
+  [[nodiscard]] const Eigen::Vector3d &vector() const { return m_vec; }
+
+  bool operator==(const Translation &other) const {
+    return m_vec == other.m_vec;
+  }
+
+  double &operator()(const int i) { return m_vec(i); }
+};
+
+template <typename T> inline const T &asEigenType(const T &obj) { return obj; };
+inline const auto &asEigenType(const Quaternion &obj) { return obj.quat(); }
+inline auto asEigenType(const Translation &obj) {
+  return Eigen::Translation<double, 3>(obj.vector());
+}
+
+template <class T_LHS, class T_RHS>
+struct combines_to_linear : std::false_type {};
+template <>
+struct combines_to_linear<Quaternion, Eigen::Matrix3d> : std::true_type {};
+template <>
+struct combines_to_linear<Eigen::Matrix3d, Quaternion> : std::true_type {};
+
+template <typename T_LHS, typename T_RHS>
+[[nodiscard]] inline std::enable_if_t<combines_to_linear<T_LHS, T_RHS>::value,
+                                      Eigen::Matrix3d>
+operator*(const T_LHS &lhs, const T_RHS &rhs) {
+  return asEigenType(lhs) * asEigenType(rhs);
+}
+
+template <class T_LHS, class T_RHS>
+struct combines_to_affine : std::false_type {};
+template <>
+struct combines_to_affine<Quaternion, Translation> : std::true_type {};
+template <>
+struct combines_to_affine<Quaternion, Eigen::Affine3d> : std::true_type {};
+template <>
+struct combines_to_affine<Eigen::Matrix3d, Translation> : std::true_type {};
+template <>
+struct combines_to_affine<Eigen::Affine3d, Quaternion> : std::true_type {};
+template <>
+struct combines_to_affine<Eigen::Affine3d, Translation> : std::true_type {};
+template <>
+struct combines_to_affine<Translation, Quaternion> : std::true_type {};
+template <>
+struct combines_to_affine<Translation, Eigen::Matrix3d> : std::true_type {};
+template <>
+struct combines_to_affine<Translation, Eigen::Affine3d> : std::true_type {};
+
+template <typename T_LHS, typename T_RHS>
+[[nodiscard]] inline std::enable_if_t<combines_to_affine<T_LHS, T_RHS>::value,
+                                      Eigen::Affine3d>
+operator*(const T_LHS &lhs, const T_RHS &rhs) {
+  return Eigen::Affine3d(asEigenType(lhs) * asEigenType(rhs));
+}
+
+template <class T_LHS> struct applies_to_vector : std::false_type {};
+template <> struct applies_to_vector<Quaternion> : std::true_type {};
+template <> struct applies_to_vector<Translation> : std::true_type {};
+
+template <typename T_LHS>
+[[nodiscard]] inline std::enable_if_t<applies_to_vector<T_LHS>::value,
+                                      Eigen::Vector3d>
+operator*(const T_LHS &lhs, const Eigen::Vector3d &rhs) {
+  return asEigenType(lhs) * rhs;
+}
```

Comment:
In this case there appear to be only 2 possible template instantiations. Wouldn't it be more readable to write these functions explicitly? Maybe even as a member of the respective classes, to help the compiler?